### DISCUSSION TITLE
4.0.3 Scripts - yet another issue compared to 3.7.x

### DISCUSSION THREAD

**Discussion by ziggekatten** [netbox-community/netbox#16405] [Help Wanted!]
Some background:

 - We run Netbox in an docker Swarm cluster since years back. Works perfectly
 - Adding an new script involves an first manual upload to get the entry into the Netbox DB
 - Further developments of scripts is done on our local PC's, and we are using pipelines that when we push to a branch, files are copied to an area mounted by our netbox and netbox worker containers

This works very well in our current  3.7.x environment. Merging and pushing, a few seconds later scripts are updated and available. Any new classes in an script file is available to be used

Now, as we are looking to future upgrades to the 4.x versions, I have discovered that this no longer works. If I add an class to an existing file, it's not available to be run in Operations or Event rules. 

Have there been changes so that also classes are represented in the DB? Is there a way of update DB with new classes? As we are running things through an pipeline with code checks, testing of APis and stuff, so using an Git data source is not really an option.

**Comment by Mailstorm-ctrl** [comment]
`pipeline with code checks, testing of APis and stuff`

Git(hub/lab/etc) can do this and more. I'm not sure where the idea of this not being possible came from.

But anyways, are you using data sources? You can use a "local" file and then when an update happens you just sync the data source and the script gets updated. I've been doing this through a git source and I can't image local is any different.
https://netboxlabs.com/docs/netbox/en/stable/features/synchronized-data/

**Comment by llamafilm** [comment]
I think Netbox manages the files in `/opt/netbox/netbox/scripts` so you are not supposed to modify those directly.  I keep my scripts in git and use a local DataSource in Netbox.  I use a git post-commit hook to run the following script on the instance.  It pulls changes and then syncs the DataSource.

```bash
REPO_DIR=/opt/myrepo
REPO_URL=https://xxxxxxxx.git

if [ -d "$REPO_DIR" ]; then
    cd $REPO_DIR
    git pull $REPO_URL
else
    git clone $REPO_URL $REPO_DIR
fi

/opt/netbox/venv/bin/python /opt/netbox/netbox/manage.py syncdatasource "myrepo"
```

**Reply by ziggekatten** [reply]
Well, we are using a lot of common python modules in subfolders etc, and using an synced datasource does not play well with that, since only scripts added to Netbox is synced.

**Reply by llamafilm** [reply]
Oh I see. Do your common modules change often? If not, maybe you’d be better off adding them to local_requirements.txt and running the upgrade.sh to install them in the venv.

**Reply by ziggekatten** [reply]
As an ISP, they change quite often as we add automation daily, and we would like to skip rebuild docker images all the time (as we used before 3.7.x)
